src/Http/Authentication.Core/src/AuthenticationMetrics.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace Microsoft.AspNetCore.Authentication;
+
+internal sealed class AuthenticationMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.Authentication";
+
+    private readonly Meter _meter;
+    private readonly Counter<long> _authenticatedRequestCount;
+    private readonly Counter<long> _challengeCount;
+    private readonly Counter<long> _forbidCount;
+    private readonly Counter<long> _signInCount;
+    private readonly Counter<long> _signOutCount;
+
+    public AuthenticationMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.Create(MeterName);
+
+        _authenticatedRequestCount = _meter.CreateCounter<long>(
+            "aspnetcore.authentication.authenticated_requests",
+            unit: "{request}",
+            description: "The total number of authenticated requests");
+
+        _challengeCount = _meter.CreateCounter<long>(
+            "aspnetcore.authentication.challenges",
+            unit: "{request}",
+            description: "The total number of times a scheme is challenged");
+
+        _forbidCount = _meter.CreateCounter<long>(
+            "aspnetcore.authentication.forbids",
+            unit: "{request}",
+            description: "The total number of times an authenticated user attempts to access a resource they are not permitted to access");
+
+        _signInCount = _meter.CreateCounter<long>(
+            "aspnetcore.authentication.sign_ins",
+            unit: "{request}",
+            description: "The total number of times a principal is signed in");
+
+        _signOutCount = _meter.CreateCounter<long>(
+            "aspnetcore.authentication.sign_outs",
+            unit: "{request}",
+            description: "The total number of times a scheme is signed out");
+    }
+
+    public void AuthenticatedRequest(string scheme, AuthenticateResult result)
+    {
+        if (_authenticatedRequestCount.Enabled)
+        {
+            var resultTagValue = result switch
+            {
+                { Succeeded: true } => "success",
+                { Failure: not null } => "failure",
+                { None: true } => "none",
+                _ => throw new UnreachableException($"Could not determine the result state of the {nameof(AuthenticateResult)}"),
```

Comment:
Is this the expected pattern? If the meter is unable to determine this, I don't think it's up to the meter to decide if an exception should be thrown or not. Hence, I wonder if simply using a value like "unknown" is a more reasonable approach here.

My main point is that the authentication stack should be responsible for determining if a certain state is valid or not, not a meter.

Reply:
I see that no guidance has been provided in documentation either. Perhaps it's something we should provide guidance for. @JamesNK, what do you think?

Reply:
The `UnreachableException` is more of an assertion than anything else. It's declaring that one of the specified conditions _must_ be true. If this exception ever gets thrown in reality, it indicates a framework bug.

I'm not sure we'd want to introduce a fourth value like `"unknown"` that customers need to be aware of.

Reply:
Maybe an alternative is to just make "none" the default case, although then we're not validating that the result actually _is_ "none".

Reply:
```suggestion
                _ => "_OTHER", // _OTHER is commonly used fallback for an extra or unexpected value. Shouldn't reach here.
```

It doesn't look like you should ever get here, but just in case, there should be a fallback value. `_OTHER` is often used: https://github.com/search?q=repo%3Aopen-telemetry%2Fsemantic-conventions%20_OTHER&type=code